### PR TITLE
Optional alpha channel support for entities

### PR DESCRIPTION
**Pull request by stujones11**
This PR allows optional use of alpha-channel for `cube`, `sprite` and `mesh` active objects. Supports dynamic updating so could be useful for temporary materialization effects, etc.

Currently entities use the `EMT_TRANSPARENT_ALPHA_CHANNEL_REF` material type flag which results in pixels being rendered either fully transparent or fully opaque depending on individual alpha level.
The `EMT_TRANSPARENT_ALPHA_CHANNEL` flag, which does support full alpha-channel, is know to have z-sorting issues, like being invisible from underwater, so cannot be simply used as the default.

This change does not apply to `upright_sprite` or `wielditem` visuals since the former already appears to support alpha (at least for the front face) and wielditems would probably look terrible with the broken depth sorting and how they are constructed, though I haven’t actually tried this.

The second commit allows for dynamic update of a mesh object’s `backface_culling` property. This is of particular importance if you wanted to use alpha support for a given player, currently player face culling is hard-coded `false` in a really crude way. Testing has shown that semi-transparent objects both look and behave better with backface culling enabled.

Using this property for players is not really recommend for the reasons pointed out above, however, it may be deemed acceptable under the right circumstances. As with nodes, this feature should be used sparingly and with caution, at least until depth sorting is improved.

Some example screenshots here: https://forum.minetest.net/viewtopic.php?f=47&t=18026#p305791

### CODE REVIEW

**Comment by sofar** [comment]
This sounds like an OK patch for now, it'll hopefully entice someone to figure out what needs to be done to make z-sorting work, and I'd rather see us open up the possibilities instead of closing them down.

**Comment by paramat** [comment]
I can't see how dynamic updating of 'backface culling' is done in your commit, it's not added as a settable object property, it looks hardcoded. I've probably missed something.
Some player skins rely on backface culling being false so keep the default unchanged.
Apart from those points the concept seems good.

**Comment by lhofhansl** [comment]
Looks good to me.

Somewhat related: If somebody can figure out why many model (not player models) have no or wrong normals that'd be great too.

**Comment by stujones11** [comment]
> I can't see how dynamic updating of 'backface culling' is done in your commit

Both material type and backface culling are set in `GenericCAO::updateTextures` which gets called every time `obj:set_properties` is used, regardless (it seems) whether the textures are updated or not.

> Note that some player skins rely on backface culling being false.

This is still the default, only now it is possible to change that if desired.

@lhofhansl I'd guess the models were exported with without normals or with errors currently unnoticeable in MT. The advanced lighting PR exposed a problem the default player model that was fixed by applying the edge-split modifier. I believe it has something to do with using quads instead of tris but don't quote me on that.

**Comment by paramat** [comment]
@sofar your blank avatar is rather annoying, you can restore your original avatar by getting it from here https://github.com/identicons/sofar.png

identicon

:3

**Comment by jastevenson303** [comment]
I like his blank identicon!

**Comment by benrob0329** [comment]
This would be really helpful, and it's ready...could this be merged please?

**Comment by paramat** [comment]
Needs review first.

File: src/object_properties.cpp [hunk 1]
```diff
@@ -164,4 +166,5 @@ void ObjectProperties::deSerialize(std::istream &is)
 	breath_max = readU16(is);
 	eye_height = readF1000(is);
 	zoom_fov = readF1000(is);
+	use_texture_alpha = readU8(is);
```
Comment (SmallJoker) [inline src/object_properties.cpp]: This will read garbage on older servers. Please check whether there's still data available or use try/catch.

Comment (stujones11) [reply src/object_properties.cpp]: Surely this won't matter unless you want to backport it?

**Comment by stujones11** [comment]
This doesn't seem to want to rebase, maybe another time then.

**Comment by stujones11** [comment]
Replaced by  #7148

File: src/object_properties.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/object_properties.cpp]: Indeed we break the backwards compatibility in this version. Never mind, the older clients will hopefully handle this without a crash.

**Pull request closed without merging**